Add Sbus Non Inverted Mode

jihlein opened the pull request:
Fixes #390.

This PR implements a non-inverted Sbus interface on the following targets: AQ32, Naze32, and Sparky.  The mode is called Sbus Non Inverted on the GCS setup screens.  The 3 targets implemented here give an F1, F3, and F4 example.

I've tested the capability to selected both Sbus and Sbus Non Inverted on all applicable ports on the above 3 targets.  Each target was able to boot and execute as expected.  As I don't have any Sbus receivers, I can not test further.  Hopefully @dustin and @Nub can test with SBus receivers.  Testing should include, at a minimum, both manual setup and wizard setup as applicable.  A quick test flight wouldn't hurt either.

As the port drop down boxes were getting long and somewhat disorganized, I alphabetized the selections so the lists don't look as disorganized.

A future PR should implement the Sbus Non Inverted mode on the remaining targets, along with the alphabetized selection list so all targets look uniform.

If there's a better way to implement some of this, let me know.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/d-ronin/dronin/403)

<!-- Reviewable:end -->

jihlein committed:
Flight: Add Sbus Non Inverted to AQ32, Naze32, and Sparky.

jihlein committed:
Flight: Add Sbus Non Inverted to AQ32, Naze32, and Sparky.

Flight: AQ32 and Naze32 updates required after rebase.

fujin commented:
jenkins, build artifacts please

tracernz commented:
This removes HoTT support on Naze?

jihlein commented:
There was no HoTT support, so I cleaned up the drop down boxes.

tracernz commented:
Looks like at least most of the code was there? Any reason not to complete it instead of removing?

jihlein commented:
I think I goofed....

jihlein committed:
Flight: Restore HoTT SUMD and HTT SUMH modes to Naze32.

jihlein commented:
HoTT support restored in Naze32.

pug398 commented:
Sparky1/X4R-SB with "A pad mod" seems fine. Wizard setup went smoothly and test flight had no issues. SBUS Non Inverted on Rx port.

jihlein commented:
@mlyle, I think we should pull this from Renatus.  I figured out what you where talking about last night and am making the changes.  Reduced the argument count to PIOS_HAL by two.  That's the good news.  Bad news is it broke a few of targets (the ones not using PIOS_HAL yet), and it's going to take some time to fix those.  Brain, Colibri, and FlyingF3 are the culprits.  The rest went smoothly.  Should we write a PR to move those targets to PIOS_HAL and hold this one until the move is complete?

@pug398, glad to hear the concept works.

jihlein committed:
Flight: PIOS_HAL rework for Sbus Non Inverted.  Brain, Colibri, and FlyingF3 broken.

jihlein commented:
Latest code pushed, not fully tested.  Brain, Colibri, and FlyingF3 do not compile.

jihlein committed:
Flight: Minimal set of changes to allow clean build of Brain, Colibri, and FlyingF3.

jihlein commented:
Minimal changes made to Brain, Colibri, and FlyingF3, everything compiles.

pug398 commented:
@jihlein Do you need a test on Brain?

jihlein commented:
@pug398, Brain is not complete, it needs to move to PIOS_HAL before it can take full advantage of the SBus Non Inverted changes.  With the minimal changes I made, it should work as before.  PIOS_HAL for the Brain is going to take some thought.

jihlein committed:
Flight: Add Sbus Non Inverted to port pull down lists for CopterControl, Quanton, RevoMini, and Sparky2 targets.

Flight: First round of review changes.

jihlein commented:
@pug38, can you rebuild with the latest files and try Sbus Non Inverted on Sparky1 again?

pug398 commented:
jenkins, build artifacts please

mlyle commented:
artifacts here
http://jar.lyle.org:8080/job/dronin/61/Nodes=winx86/

Sorry I'm abusing jenkins at the moment

pug398 commented:
@jihlein quick setup and indoor flight. Everything still looks good. There is about a 10 second delay in the detection of stick movement on the max limit page but don't think that is related to this PR.

mlyle commented:
jenkins, build artifacts please

mlyle commented:
@pug398 -- did you test noninverted or inverted (or both?)

pug398 commented:
Non inverted and getting ready to test inverted.

pug398 commented:
Both doing it. It is late and I may be crazy but there is text about moving all controls and switches to max travel. At first the text takes up two lines but somewhere after a delay it blinks and text only takes up one line. At this point sticks will register. I also got a reset at this point a couple of times. Will check more tomorrow...

mlyle commented:
Hm.  I did make a UI change to force you to get a minimal range on all channels before allowing you to advance past the max travel screen.  The "next" thing is greyed until all configured channels (including switches) have moved at least a minimum amount.  

That doesn't sound good-- especially the resets.  What TX/RX are you using?

I will try to get to testing input wizard again.

mlyle commented:
Review status: 0 of 38 files reviewed at latest revision, 6 unresolved discussions.

---

*[flight/PiOS/Common/pios_hal.c, line 385 \[r5\]](https://reviewable.io:443/reviews/d-ronin/dronin/403#-K7YN-2WmnpOEw3KVoAa:-K7YN-2WmnpOEw3KVoAb:1548291931) ([raw file](https://github.com/d-ronin/dronin/blob/a688510a6ac6a9be88880bb644d897a18bc3e474/flight/PiOS/Common/pios_hal.c#L385)):*
This needs to not be a pointer, because as a pointer it often points to non-writable flash.  It should be struct pios_usart_cfg pios_hal_usart_temp_cfg;  Then the assignment can be done, I believe, with pios_hal_usart_temp_cfg = *usart_port_cfg.

---

*[flight/PiOS/Common/pios_hal.c, line 492 \[r5\]](https://reviewable.io:443/reviews/d-ronin/dronin/403#-K7YNk42S26fh_-VXDFo:-K7YNk42S26fh_-VXDFp:-1401440317) ([raw file](https://github.com/d-ronin/dronin/blob/a688510a6ac6a9be88880bb644d897a18bc3e474/flight/PiOS/Common/pios_hal.c#L492)):*
I think PIOS_USART_Init needs to be changed to not save a pointer to the config-- instead just holding the pointer to the regs.  Thankfully, the regs is already a pointer to a separate structure so it is not "copied" by the proposed assignment.

This is because pios_hal_usart_temp_cfg is on stack and only has scope/life as long until PIOS_HAL_ConfigurePort returns.

:/  Which is becoming a much more substantial change, because it breaks ChangeBaud as written, because it wants to see the init stuff later. 

---

*[flight/targets/naze32/board-info/board_hw_defs.c, line 724 \[r5\]](https://reviewable.io:443/reviews/d-ronin/dronin/403#-K7YQ9GbPfJxwwK_LvqS:-K7YQ9Gcm-Z8f5c0hbtJ:-583402638) ([raw file](https://github.com/d-ronin/dronin/blob/83469bb150cb8415010c85a368b6bee862f1e61c/flight/targets/naze32/board-info/board_hw_defs.c#L724)):*
I see you've de-constified this so it works.  Unfortunately this moves all this stuff to ram which really increases the pressure on F1 targets.

---

*[flight/targets/naze32/board-info/board_hw_defs.c, line 760 \[r5\]](https://reviewable.io:443/reviews/d-ronin/dronin/403#-K7YQ5ZqWsTlNBX_iD6v:-K7YQ5ZqWsTlNBX_iD6w:-1114765562) ([raw file](https://github.com/d-ronin/dronin/blob/83469bb150cb8415010c85a368b6bee862f1e61c/flight/targets/naze32/board-info/board_hw_defs.c#L760)):*
Probably shouldn't be renamed?

---

*[flight/targets/naze32/board-info/pios_board.h, line 157 \[r5\]](https://reviewable.io:443/reviews/d-ronin/dronin/403#-K7YQIzP5TVgOxjzBIM8:-K7YQIzP5TVgOxjzBIM9:812747157) ([raw file](https://github.com/d-ronin/dronin/blob/a688510a6ac6a9be88880bb644d897a18bc3e474/flight/targets/naze32/board-info/pios_board.h#L157)):*
Whitespace-- converted to space from tab.. should be consistent with below/above

---

*[flight/targets/pipxtreme/fw/pios_board.c, line 164 \[r5\]](https://reviewable.io:443/reviews/d-ronin/dronin/403#-K7YQSIh5Tftf8QnJ9Vh:-K7YQSIh5Tftf8QnJ9Vi:-1787918134) ([raw file](https://github.com/d-ronin/dronin/blob/a688510a6ac6a9be88880bb644d897a18bc3e474/flight/targets/pipxtreme/fw/pios_board.c#L164)):*
space at end of line

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/d-ronin/dronin/403)*
<!-- Sent from Reviewable.io -->

mlyle commented:
I'm sorry this is proving a pain.  i didn't fully understand what you did when I started reviewing.  This is much closer, but I'd rather not move the configuration structures to RAM if it can be avoided.  Unfortunately, that takes some degree of refactoring of the pios_usart implementations.

mlyle commented:
No choice but to defer this, with other churn and concern about sbus maturity in general.  Will send a PR to this in a few days fixing the usart half.

jihlein commented:
After ignoring this for a day or so, I believe I have the usart refactoring figured out after taking another look at it.  Will test my changes over the next day or so and let you know.

jihlein committed:
Flight: Usart refactor, only AQ32, Naze32, and Sparky compile at this point.

jihlein committed:
Flight: CopterControl, Pipxtreme, Quanton, RevoMini, and Sparky2 now compile with the refactored USART code.

jihlein commented:
Note: Waiting for #411, #413, #414, and #444 to merge.

jihlein committed:
Flight: Updates to allow Colibri to compile.

jihlein committed:
Flight: Updates to allow DiscoveryF4 to compile.

jihlein committed:
Flight: Updates to allow FlyingF3 to compile.

jihlein committed:
Flight: Updates to allow Lux to compile, add sbus non inverted.

jihlein commented:
All targets but Brain now compiling.

jihlein committed:
Flight: Last file cleanups, all targets build now.

Flight: Formatting cleanup.

Flight: Correct rebase errors.

Flight: Fix rebase error in FlyingF3.

jihlein commented:
Removed WIP from title.  All targets building.  Ready for testing and review.

Let the fun begin.....

mlyle commented:
Not going to mess with this for a few days until I feel a little better about next, though.

mlyle commented on flight/PiOS/Common/pios_hal.c:
```diff
@@ -379,17 +382,144 @@ void PIOS_HAL_ConfigurePort(HwSharedPortTypesOptions port_type,
+#if defined(PIOS_INCLUDE_DSM)
+		if (dsm_cfg && usart_port_cfg) {
+			usart_port_params.init.USART_BaudRate              = 115200;
+			//usart_port_params.init.USART_WordLength          = USART_WordLength_8b;
```
Remove commented code

mlyle commented on ground/gcs/src/plugins/coreplugin/iboardtype.h:
```diff
@@ -163,6 +163,7 @@ class CORE_EXPORT IBoardType : public QObject
         INPUT_TYPE_PPM,
         INPUT_TYPE_DSM,
         INPUT_TYPE_SBUS,
+		INPUT_TYPE_SBUSNONINVERTED,
```
Indentation (need to match the wrong indentation on other lines ):

tracernz replied:
We use spaces in GCS code.

mlyle commented:
Looks very nice.  I owe you a more thorough read.  net savings of 2700 lines is very nice :)

jihlein commented:
Regarding the commented code in PIOS_HAL.c, no problem removing it but it was intentionally left as documentation to what the complete port setting would be.  That would be the only reason to consider leaving it in place.

mlyle commented:
Alternatively, uncomment; optimizer should take it out.

jihlein commented:
If that's the case, I'd prefer to uncomment it, leaving it there for documentation's sake.

mlyle commented:
Also needs a rebase unfortunately; probably the revo changes.

jihlein commented:
I loaded this on a Sparky, did a quick test of external mag setup on the flex port, port configured correctly.  

I'll try to get this loaded on my AQ32 tricopter in the near future.  That will cover gps config, msp config, openlog config, external mag config, FrSky sensor hub config, and ppm config.

Is there someone who could run a test of sbus non inverted?  Any target will do.

dustin commented:
I could do something on a desk.  I've got radios that do SBNI

On Thu, Feb 4, 2016 at 5:33 PM John Ihlein notifications@github.com wrote:

> I loaded this on a Sparky, did a quick test of external mag setup on the
> flex port, port configured correctly.
> 
> I'll try to get this loaded on my AQ32 tricopter in the near future. That
> will cover gps config, msp config, openlog config, external mag config,
> FrSky sensor hub config, and ppm config.
> 
> Is there someone who could run a test of sbus non inverted? Any target
> will do.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/d-ronin/dRonin/pull/403#issuecomment-180143675.

mlyle commented:
I will try to get to it soon.  I was hoping to merge it to my "kitchensink" branch, but there's some conflicts with the things I'm doing with pipxtreme so it should probably wait to the next round of test (but possibly merge first).

dustin commented:
Looking to do some testing here.  I started by attempting to rebase it over next.  It's got a bunch of trailing whitespace, which rebase complained about, but I'm not entirely sure if that caused the issues, but it's a thing to resolve.

(testing it as-is)

---

Review status: 0 of 52 files reviewed at latest revision, all discussions resolved.

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/d-ronin/dronin/403#-:-K9opXf7L27ZHcYnORC_:-1648984712)_

<!-- Sent from Reviewable.io -->

jihlein commented:
I just finished rebasing it.  Did a bunch of testing with AQ32 today.  Here's what I verified:

1)PPM rx setup
2)GPS operation
3)External mag operation and orientation
4)FrSky sensor hub operation
5)MSP operation (note a bug with stack sizes was found here and corrected on next)
6)VCP pass thru with a baud rate change to 115200 from 57600 (used to setup MWOSD)
7)OpenLog operation

All ports setup and functioned as expected.  Tests of Sbus and Sbus non inverted would be helpful, as would testing on an F1 and F3 target.  I've done some basic tests on Sparky.

I've got a little more setup work to do, but hope to fly the AQ32 target tomorrow.

mlyle commented:
```
/var/lib/jenkins/jobs/dron/workspace/Nodes/master/build/fw_flyingf3/logging.o: In function `LoggingInitialize':
/var/lib/jenkins/jobs/dron/workspace/Nodes/master/flight/Modules/Logging/logging.c:149: undefined reference to `pios_com_openlog_logging_id'
/var/lib/jenkins/jobs/dron/workspace/Nodes/master/build/fw_flyingf3/pios_hal.o: In function `PIOS_HAL_ConfigurePort':
/var/lib/jenkins/jobs/dron/workspace/Nodes/master/flight/PiOS/Common/pios_hal.c:655: undefined reference to `pios_com_openlog_logging_id'
collect2: error: ld returned 1 exit status
```

mlyle commented:
Looks like a really clean/nice history btw.

dustin commented:
I couldn't get my x6r working on it with a sparky 1.  It worked OK on next.

I also failed to get my 4ch remix working in either direction.  It didn't work on next either, though, so I might just be doing it wrong.  heh.

---

Review status: 0 of 52 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/d-ronin/dronin/403#-:-K9p6oYKXVb9Lg_jMh4O:1391826377)_

<!-- Sent from Reviewable.io -->

pug398 commented:
jenkins, build artifacts please

mlyle commented:
@jihlein it looks like there's something wrong with this rebase-- did you base off upstream/next?   There's alreayd-commited PRs in the list.

@pug398 -- the build failed, it looks like.  Also I'm not sure if this is worth testing yet, as it wasn't working last night.

jihlein committed:
Flight: Add receiver usart configuration to Sparky.

Flight: Rebase file cleanups.

jihlein commented:
Rebased to current next.

Flew the AQ32 setup on a tricopter as described a few comments above with no issues, all ports setting up correctly and flew with no surprises.  Pulled telemetry off OpenLog and parse cleanly with Octave.

Setup a Sparky with DSM satellite on RvcrPort, also on a tricopter.  Everything setup correctly and flew with no issues.

pug398 commented:
@jihlein I have SB/SBNI rx mounted on Brain quad if you still need another target tested.

jihlein commented:
@pug398, that would be perfect, the sb/sbni still need some testing/verification.  On the Brain FC, even better.

mlyle commented:
I believe @dustin tried this and I don't think it worked (either non-inverted or inverted).  I have desk-checked it and didn't see anything obviously wrong, but have not gotten a chance to try it myself.

mlyle commented:
Man, I read the code again and it sure looks like it should work.  If it's broken, it's something subtle and dumb :P

jihlein commented:
I believe it was tried on Sparky, and there was a problem that was corrected with the last commit.

pug398 commented:
jenkins, build artifacts please

pug398 commented:
Ok so this did not pull in old settings and brain needed to be fully erased. With 90 changes I guess that can be expected :) With a settings import it flew fine on sbus and moving rx to sbni output it flew fine with just a main port change to sbus non inverted. A full erase and setup using radio wizard needs to be done as that is where issues usually show up. Will try to get to that tomorrow afternoon. Looking good so far.

jihlein commented:
Very cool @pug398, that's great to hear. Thanks!

tracernz commented:
@mlyle Jenkins didn't call back with the artifacts but it has finished building them. Does it not do that with merge conflicts?

mlyle commented:
It's supposed to and the comment script run, but it apparently didn't happen.  It just uses the pull id, which was set right in the build.  So maybe there was a github glitch or something (I don't check success status or retry failures).

pug398 commented:
On a fresh Brain setup using sbus non inverted, input setup via radio wizard completed without any problems. Subsequent flights went well.

jihlein commented:
Not sure why, but the make ef_target cmds are are failing.

recipe for target 'c:/dronin/build/ef_brain/ef_brain.fw.pre.pad' failed, for example.  AQ32 and Sparky fail the same way.

tracernz commented:
@jihlein Can't replicate that on Linux. Can you open an issue with further details?

jihlein committed:
Flight: Allow for Uart single wire mode on F4 targets.

Flight: Rebase correction.

jihlein commented:
Rebased to current next earlier today.  Here's the current test matrix status.

![image](https://cloud.githubusercontent.com/assets/3869782/13201025/e78ead96-d81c-11e5-9618-cd26b0a284b9.png)

X1: AQ32 test of combridge also changed the default baud rate of 57600 to 115200 to talk to and setup an MWOSD, effectively checking the changes to PIOS_USART_ChangeBaud on the F4 platform.

X2: Sparky test of  telemetry also changed the default baud rate of 57600 to 115200 to talk to bluetooth radio, effectively checking the changes of PIOS_USART_ChangeBaud on the F3 platform.

mlyle commented:
Would be nice to see sbus/sbusni work on F3.  Also FrskySport is different on F3 too.

pug398 commented:
sbus/sbusni flight tested on sparky1 no problems. FrskySport tested on sparky1 main port with basic telemetry displayed on Taranis. No problems indicated.

pug398 commented:
Bluetooth telemetry tested ok on sparky flexi port @115k as well.

mlyle commented:
I have skimmed this and it looks good to me-- I'm getting a little blind to it from reading it several times.

It has done well on all the key tests.

If there's no objections I'm inclined to merge this and shake it out the rest of the way on next.

tracernz commented:
Reviewed 1 of 20 files at r3, 1 of 4 files at r5, 18 of 50 files at r6, 6 of 16 files at r7, 18 of 25 files at r8, 6 of 6 files at r9, 2 of 2 files at r10.
Review status: all files reviewed at latest revision, all discussions resolved.

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/d-ronin/dronin/403)_

<!-- Sent from Reviewable.io -->

jihlein commented:
Updated test matrix.

tracernz commented:
S.BUS on F1 would be good if anyone has some hardware available to test.

mlyle commented:
I don't expect F1 will be different than F4-- it works the same all in all, doesn't it?

We're accumulating a fair bit of test debt and minor regression, so it'll be time to shake things out thoroughly soon.  Best if it happens together IMO.
